src/operator/join.rs

```diff
@@ -0,0 +1,223 @@
+use crate::{
+    algebra::{finite_map::KeyProperties, FiniteMap, MapBuilder, ZRingValue},
+    circuit::{
+        operator_traits::{BinaryOperator, Operator},
+        Circuit, Scope, Stream,
+    },
+};
+use std::{borrow::Cow, marker::PhantomData};
+
+impl<P, I1> Stream<Circuit<P>, I1>
+where
+    P: Clone + 'static,
+{
+    /// Apply [`Join`] operator to `self` and `other`.
+    pub fn join<K, V1, V2, V, W, F, I2, Z1, Z2, Z>(
+        &self,
+        other: &Stream<Circuit<P>, I2>,
+        f: F,
+    ) -> Option<Stream<Circuit<P>, Z>>
+    where
+        K: KeyProperties,
+        V1: KeyProperties,
+        V2: KeyProperties,
+        W: ZRingValue,
+        I1: FiniteMap<K, Z1>,
+        for<'a> &'a I1: IntoIterator<Item = (&'a K, &'a Z1)>,
+        for<'a> &'a Z1: IntoIterator<Item = (&'a V1, &'a W)>,
+        Z1: 'static,
+        I2: FiniteMap<K, Z2>,
+        for<'a> &'a I2: IntoIterator<Item = (&'a K, &'a Z2)>,
+        for<'a> &'a Z2: IntoIterator<Item = (&'a V2, &'a W)>,
+        Z2: 'static,
+        F: Fn(&K, &V1, &V2) -> V + 'static,
+        V: 'static,
+        Z: Clone + MapBuilder<V, W> + 'static,
+    {
+        self.circuit()
+            .add_binary_operator(Join::new(f), self, other)
+    }
+}
+
+/// Join two indexed Z-sets.
+///
+/// The operator takes two streams of indexed Z-sets and outputs
+/// a stream obtained by joining each pair of inputs.
+///
+/// An indexed Z-set is a map from keys to a Z-set of values associated
+/// with each key.  Both input streams must use the same key type `K`.
+/// Indexed Z-sets are produced for example by the
+/// [`Index`](`crate::operator::Index`) operator.
+///
+/// # Type arguments
+///
+/// * `K` - key type.
+/// * `V1` - value type in the first input stream.
+/// * `V2` - value type in the second input stream.
+/// * `V` - value type in the output Z-set.
+/// * `W` - weight type used by both inputs.
+/// * `F` - join function type: maps key and a pair of values from input Z-sets
+///   to an output value.
+/// * `I1` - indexed Z-set type in the first input stream.
+/// * `I1` - indexed Z-set type in the second input stream.
+/// * `Z1` - value Z-set type in the first input stream.
+/// * `Z2` - value Z-set type in the second input stream.
+/// * `Z` - output Z-set type.
```

Comment:
This *screams* "traits with associated types" to me

Reply:
I am not a fan of this signature either, but I don't see how to use associated types here without restricting the set of supported input/output types.  Any suggestions?

Reply:
On a second thought, I think there's a way. I'll submit a revised version of this shortly.

Reply:
Ok, I got rid of two of the type arguments (`Z1` and `Z2`) and am open for suggestions to improve things further.

Reply:
Maybe make the index type a param on a trait for the input streams? That seems like it should be fairly related, but I'm also not sure how that may impact varying index types (may end up requiring noop wrapper types to change index types or something).

Although on the *inverse*, the index type could make the input stream type a parameter on itself a la
```rust
trait Index
    type Value;
    type Weight;
}
```

Reply:
The inverse does not feel right, `IndexedZSet` does not need to know about circuits or streams.  A trait for streams with index as an associated types might work.  I'll try that.

Reply:
Meh, it's doable (in fact either way, as you suggested), but feels like bending backward.  Sure, we can reduce the number of generic arguments, but I'm not sure it gets simpler overall.